ambry-server/src/main/java/com/github/ambry/server/StatsManager.java

```diff
@@ -435,53 +478,108 @@ long getPermanentDeleteTombstoneSize() {
   }
 
   /**
-   * Runnable class that collects, aggregate and publish stats via methods in StatsManager.
+   * Runnable class that collects, publish account stats to mysql database and local backup file.
    */
-  private class StatsAggregator implements Runnable {
+  class AccountStatsPublisher implements Runnable {
     private volatile boolean cancelled = false;
     private final AccountStatsMySqlStore accountStatsMySqlStore;
 
-    StatsAggregator(AccountStatsMySqlStore accountStatsMySqlStore) {
+    AccountStatsPublisher(AccountStatsMySqlStore accountStatsMySqlStore) {
       this.accountStatsMySqlStore = accountStatsMySqlStore;
     }
 
     @Override
     public void run() {
-      logger.info("Aggregating stats for local report");
-      try {
-        resetDeleteTombstoneStats();
-        long totalFetchAndAggregateStartTimeMs = time.milliseconds();
-        StatsSnapshot aggregatedSnapshot = new StatsSnapshot(0L, new HashMap<>());
-        List<PartitionId> unreachablePartitions = new ArrayList<>();
-        Iterator<PartitionId> iterator = (new HashSet<>(partitionToReplicaMap.keySet())).iterator();
-        while (!cancelled && iterator.hasNext()) {
-          PartitionId partitionId = iterator.next();
-          logger.info("Aggregating stats for local report started for store {}", partitionId);
-          collectAndAggregate(aggregatedSnapshot, partitionId, unreachablePartitions);
-        }
-        aggregatedSnapshot.updateValue();
-        List<String> unreachableStores = examineUnreachablePartitions(unreachablePartitions);
-        updateAggregatedDeleteTombstoneStats();
-        metrics.initDeleteStatsGaugesIfNeeded();
-        if (!cancelled) {
-          metrics.totalFetchAndAggregateTimeMs.update(time.milliseconds() - totalFetchAndAggregateStartTimeMs);
-          StatsHeader statsHeader = new StatsHeader(StatsHeader.StatsDescription.STORED_DATA_SIZE, time.milliseconds(),
-              partitionToReplicaMap.keySet().size(), partitionToReplicaMap.keySet().size() - unreachableStores.size(),
-              unreachableStores);
-          // First write to account stats
-          StatsWrapper statsWrapper = new StatsWrapper(statsHeader, aggregatedSnapshot);
+      Object lock = accountStatsMySqlStore != null ? accountStatsMySqlStore : this;
+      synchronized (lock) {
+        logger.info("Aggregating account stats for local report");
+        try {
+          resetDeleteTombstoneStats();
```

Comment:
Could you add a few more inline comments explaining what each of the steps does?  It's a lot of code with relatively few comments.

Reply:
updated.